File: 3544_subtree_inversion_sum/Solution.kt
// LeetCode 3544 - Subtree Inversion Sum
// https://leetcode.com/problems/subtree-inversion-sum/

class Solution {
    lateinit var graph: Array<ArrayList<Int>>
    lateinit var nums: IntArray
    lateinit var parent: IntArray
    var k = 0
    lateinit var memo: HashMap<String, Long>

    fun subtreeInversionSum(edges: Array<IntArray>, nums: IntArray, k: Int): Long {
        val n = edges.size + 1
        this.nums = nums
        this.k = k
        graph = Array(n) { ArrayList() }
        for (e in edges) {
            graph[e[0]].add(e[1])
            graph[e[1]].add(e[0])
        }
        parent = IntArray(n) { -1 }
        memo = HashMap()
        return dp(0, k, false)
    }

    fun dp(u: Int, steps: Int, inv: Boolean): Long {
        val key = "$u,$steps,$inv"
        memo[key]?.let { return it }
        var num = nums[u].toLong()
        if (inv) num = -num
        var negNum = -num
        for (v in graph[u]) {
            if (v == parent[u]) continue
            parent[v] = u
            var ns = steps + 1
            if (ns > k) ns = k
            num += dp(v, ns, inv)
            if (steps == k) negNum += dp(v, 1, !inv)
        }
        var res = num
        if (steps == k && negNum > res) res = negNum
        memo[key] = res
        return res
    }
}

File: 3544_subtree_inversion_sum/solution.ts
// LeetCode 3544 - Subtree Inversion Sum
// https://leetcode.com/problems/subtree-inversion-sum/

export function subtreeInversionSum(edges: any, nums: any, k: any): any {
    const n = edges.length + 1;
    const graph = Array.from({length: n}, () => []);
    for (const e of edges) {
        graph[e[0]].push(e[1]);
        graph[e[1]].push(e[0]);
    }
    const parent = new Array(n).fill(-1);
    const memo = new Map();
    function dp(u: any, steps: any, inv: any): any {
        const key = u + ',' + steps + ',' + inv;
        if (memo.has(key)) return memo.get(key);
        let num = nums[u];
        if (inv) num = -num;
        let negNum = -num;
        for (const v of graph[u]) {
            if (v === parent[u]) continue;
            parent[v] = u;
            let ns = steps + 1;
            if (ns > k) ns = k;
            num += dp(v, ns, inv);
            if (steps === k) negNum += dp(v, 1, !inv);
        }
        let res = num;
        if (steps === k && negNum > res) res = negNum;
        memo.set(key, res);
        return res;
    }    return dp(0, k, false);
}
